crates/turbopath/src/relative_unix_path_buf.rs

```diff
@@ -1,75 +1,171 @@
-use std::path::PathBuf;
+use std::{fmt::Debug, io::Write};
 
-use serde::Serialize;
+use bstr::{BString, ByteSlice};
 
-use crate::{IntoUnix, PathValidationError};
+use crate::{PathError, PathValidationError};
 
-#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Serialize)]
-pub struct RelativeUnixPathBuf(PathBuf);
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct RelativeUnixPathBuf(BString);
 
 impl RelativeUnixPathBuf {
-    /// Create a new RelativeUnixPathBuf from a PathBuf by calling `into_unix()`
-    ///
-    /// NOTE: `into_unix` *only* converts Windows paths to Unix paths *on* a
-    /// Windows system. Do not pass a Windows path on a Unix system and
-    /// assume it'll be converted.
-    ///
-    /// # Arguments
-    ///
-    /// * `path`:
-    ///
-    /// returns: Result<RelativeUnixPathBuf, PathValidationError>
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// ```
-    pub fn new(path: impl Into<PathBuf>) -> Result<Self, PathValidationError> {
-        let path = path.into();
-        if path.is_absolute() {
-            return Err(PathValidationError::NotRelative(path));
+    pub fn new(path: impl Into<Vec<u8>>) -> Result<Self, PathError> {
+        let bytes: Vec<u8> = path.into();
+        if bytes.first() == Some(&b'/') {
+            return Err(PathValidationError::not_relative_error(&bytes).into());
         }
-
-        Ok(RelativeUnixPathBuf(path.into_unix()?))
+        Ok(Self(BString::new(bytes)))
     }
 
-    pub fn to_str(&self) -> Result<&str, PathValidationError> {
-        self.0
+    pub fn as_str(&self) -> Result<&str, PathError> {
+        let s = self
+            .0
             .to_str()
-            .ok_or_else(|| PathValidationError::InvalidUnicode(self.0.clone()))
+            .or_else(|_| Err(PathError::Utf8Error(self.0.as_bytes().to_owned())))?;
+        Ok(s)
+    }
+
+    // write_escaped_bytes writes this path to the given writer in the form
+    // "<escaped path>", where escaped_path is the path with '"' and '\n'
+    // characters escaped with '\'.
+    pub fn write_escapted_bytes<W: Write>(&self, writer: &mut W) -> Result<(), PathError> {
+        writer.write_all(&[b'\"'])?;
+        // i is our pointer into self.0, and to_escape_index is a pointer to the next
+        // byte to be escaped. Each time we find a byte to be escaped, we write
+        // out everything from i to to_escape_index, then the escape byte, '\\',
+        // then the byte-to-be-escaped. Finally we set i to 1 + to_escape_index
+        // to move our pointer past the byte we just escaped.
+        let mut i: usize = 0;
+        while i < self.0.len() {
+            if let Some(mut to_escape_index) = self.0[i..]
```

Comment:
I think the `.enumerate()` call is in the wrong spot, it will enumerate the bytes to be escaped, not the underlying bytes. Do you know if fixing that will cause a tuple to be allocated and passed through the filter function?

Reply:
Ah yeah the enumerate should be before. I wouldn't worry about the tuple allocation because tuples are value types so they'd be on the stack, and Rust is probably smart enough to split that into two registers anyways.

Reply:
Implemented a version of this.